Add subscriptions reference

hillaryfraley opened the pull request:
## Description
Adds a subscriptions reference doc using content from the agent and check references, notes from discussion with Todd and Caleb, and Todd's demo.

## Motivation and Context
Closes https://github.com/sensu/sensu-docs/issues/2679
Supports the observability pipeline docs IA work.

hillaryfraley committed:
Add subscriptions reference

cwjohnston commented on content/sensu-go/6.1/observability-pipeline/observe-schedule/subscriptions.md:
```diff
@@ -0,0 +1,123 @@
+At the core of this model are Sensu agent and check subscriptions, which are equivalent to "topics" in a traditional publish/subscribe system.
+Sensu entities become subscribers to these "topics" via their individual `subscriptions` attributes.
+Sensu checks also have a `subscriptions` attribute, where you specify which subscribers will execute the checks.
+The Sensu backend schedules checks and publishes check execution requests to entities.
```
```suggestion
The Sensu backend schedules checks and publishes check execution requests to entities.
```
 
I think this sentence needs to better describe the connection between check subscriptions, entity subscriptions and the scheduler.



From a later passage:

> For Sensu to execute a check, the check definition must include a subscription that matches the subscription of at least one Sensu agent.

This is an important concept which isn't clear in the first paragraph.

My mental model for subscriptions is a venn diagram. We should consider adding one here or in the examples below.

hillaryfraley replied:
[Here's a draft Venn diagram](https://app.lucidchart.com/invitations/accept/77c4fd1f-e467-41c2-af76-48ca32b47645) I put together. To my eyes, this is not polished enough to go in the docs, but I am going for accuracy before beauty right now. What do you think?

cwjohnston replied:
I think this is a good starting point! I have given what I hope are self-explanatory names to some of the checks you had laid out, do those changes make sense? 

Would the diagram feel more complete if the names of the subscriptions fit inside the overlapping areas? I'm not sure how to do that in a visually appealing way.

cwjohnston commented on content/sensu-go/6.1/observability-pipeline/observe-schedule/subscriptions.md:
```diff
@@ -0,0 +1,123 @@
+Sensu checks also have a `subscriptions` attribute, where you specify which subscribers will execute the checks.
+The Sensu backend schedules checks and publishes check execution requests to entities.
+
+Subscriptions are powerful service-based monitoring primitives that allow you to effectively monitor for specific behaviors or characteristics that correspond to the function provided by a particular system.
```
```suggestion
Subscriptions are powerful service-based monitoring primitives that allow you to effectively monitor for specific behaviors or characteristics that correspond to the function provided by a particular system.
```

This statement is true, but really doing much for me. Either here or earlier in the document, we need to better explain how the loose coupling used by publish/subscribe model -- and therefore subscriptions -- are beneficial. 

A few thoughts:

Publish/subscribe is important for ephemeral or elastic infrastructure where the names of things and number of things change over time.

Using a publish/subscribe model, humans writing checks don't need to know the specific names of the entities that will run the check.

Conversely, entities do not need to know the specific check names they should execute. 

The backend correlates overlapping subscriptions between checks and entities to determine which entities should receive execution requests for a given check.

cwjohnston replied:
Maybe part of what I'm missing here is enough explanation of pub/sub that an unfamiliar reader will understand the connection to the next sentence?

cwjohnston reviewed (changes requested):
This is headed in the right direction! With some revisions it to make a stronger sales pitch for pub/sub, this will be helpful to our audience!

hillaryfraley committed:
Updates based on Cameron's feedback

hillaryfraley committed:
Merge branch 'master' into subscriptions

hillaryfraley commented:
@cwjohnston thank you. I felt like I was saying the same things without really explaining. Your comments helped me figure out what I should change to make this more helpful.

I added a draft Venn diagram in a previous comment for you to take a look at, and updated the subscriptions reference draft. I think this is ready for another look.

cwjohnston commented on content/sensu-go/6.1/observability-pipeline/observe-schedule/subscriptions.md:
```diff
@@ -0,0 +1,130 @@
+In other words, subscriptions are configured for both checks and agents:
+
+- To configure subscriptions for a check, add one or more subscription names in the [check `subscriptions` attribute][15].
+- To configure subscriptions for an agent, set the [`subscriptions` flag][2] and specify the subscriptions that include the checks the agent's entities should execute.
```
```suggestion
- To configure subscriptions for an agent, configure the [`subscriptions`][2] by specifying the subscriptions that include the checks the agent's entities should execute.
```

cwjohnston commented on content/sensu-go/6.1/observability-pipeline/observe-schedule/subscriptions.md:
```diff
@@ -0,0 +1,130 @@
+- To configure subscriptions for a check, add one or more subscription names in the [check `subscriptions` attribute][15].
+- To configure subscriptions for an agent, set the [`subscriptions` flag][2] and specify the subscriptions that include the checks the agent's entities should execute.
+
+You must also set the `publish` attribute to `true` in the check definition.
```
It is possible to manually schedule execution of a check when publish is `false`. I believe such a manually requested execution can target the subscriptions defined in the check, or an alternate list of subscriptions. I think this bears mentioning as well.

hillaryfraley committed:
Apply suggestion from code review

Co-authored-by: Cameron Johnston <cameron@sensu.io>

hillaryfraley committed:
Update ad hoc text and add diagram

hillaryfraley commented:
@cwjohnston I added the info about ad hoc scheduling and a cleaned-up Venn diagram. I think this might be ready to publish -- wdyt?

cwjohnston commented on content/sensu-go/6.1/observability-pipeline/observe-schedule/subscriptions.md:
```diff
@@ -0,0 +1,158 @@
+However, only Entity A and Entity C will execute `checkSshdProcess` &mdash; Entity B does not include the `linux` subscription required to execute `checkSshdProcess`.
+
+{{< figure src="/images/subscriptions_venn.png" alt="Venn diagram example of Sensu check execution based on subscriptions" link="/images/subscriptions_venn.png" target="_blank" >}}
+<!-- Diagram source: https://app.lucidchart.com/invitations/accept/0aa33f39-128d-449d-8312-04f415d8cfca -->
```
```suggestion
<!-- Diagram source: https://app.lucidchart.com/documents/view/4ed31993-de03-46ed-92b9-0d9eed176439 -->
```

I think we should be using links to `documents` in LucidChart. The links with `invitations/accept` might grant access to documents which we don't want to provide to just anybody.

Then again, I can't tell from the link whether the invitation is simply to view the document, which would be a non-issue.

hillaryfraley replied:
I don't know why or when this changed, but I can't seem to generate a LucidChart link that doesn't include "invitations/accept" anymore. The links I'm generating are all restricted to Sensu though.

cwjohnston commented on content/sensu-go/6.1/observability-pipeline/observe-schedule/subscriptions.md:
```diff
@@ -0,0 +1,158 @@
+
+{{< code shell >}}
+{
+  "check": "check-cpu",
```
```suggestion
  "check": "check_cpu",
```

cwjohnston commented on content/sensu-go/6.1/observability-pipeline/observe-schedule/subscriptions.md:
```diff
@@ -0,0 +1,158 @@
+If you want Sensu to automatically schedule and execute a check according to its subscriptions, set the [`publish` attribute][12] to `true` in the check definition.
+
+You can also manually schedule [ad hoc check execution][11] with the [check API][16], whether the `publish` attribute is set to `true` or `false`.
+To target the subscriptions defined in the check, include only the check name in the request body (e.g. `"check": "check-cpu"`).
```
```suggestion
To target the subscriptions defined in the check, include only the check name in the request body (e.g. `"check": "check_cpu"`).
```

Suggest renaming `check-cpu` to `check_cpu` for consistency with check name used in the above diagram.
Perhaps consider renaming checks in the diagram to use consistent `snake_case` or `camelCase` names?

cwjohnston reviewed (approved):
I think this looks good! 👍  I recommend getting a review from engineering for technical correctness before we ship it.

hillaryfraley committed:
Update to use check_cpu

Co-authored-by: Cameron Johnston <cameron@sensu.io>

hillaryfraley committed:
Update check names in diagram and text

hillaryfraley commented:
Suggestion from Cameron: compare subscriptions to a role in config mgmt tool

ccressent commented on content/sensu-go/6.1/observability-pipeline/observe-schedule/subscriptions.md:
```diff
@@ -0,0 +1,158 @@
+However, only Entity A and Entity C will execute `check_sshd_process` &mdash; Entity B does not include the `linux` subscription required to execute `check_sshd_process`.
+
+{{< figure src="/images/subscriptions_venn.png" alt="Venn diagram example of Sensu check execution based on subscriptions" link="/images/subscriptions_venn.png" target="_blank" >}}
+<!-- Diagram source: https://app.lucidchart.com/lucidchart/invitations/accept/25d40e7e-7293-4189-84e0-ef8c3a4c8ec5 -->
```
Nitpick/Idea: I managed to grok the Venn diagram easily enough, but I wonder
if a different representation might help? A table with the subscriptions for
each entity and check, followed by trees/graphs showing the infered mapping
between checks and entities?

hillaryfraley replied:
I agree there is probably a better way to represent these connections. I'll create an issue to improve this diagram.

ccressent commented on content/sensu-go/6.1/observability-pipeline/observe-schedule/subscriptions.md:
```diff
@@ -0,0 +1,158 @@
+
+## Configure subscriptions
+
+For Sensu to execute a check, the check definition must include a subscription that matches a subscription for at least one Sensu entity.
```
Not sure if this is intentional or not, but this is sentence is an exact copy of a sentence 1 or 2 paragraphs above.

hillaryfraley replied:
I did leave it that way intentionally, but I can rephrase it so it's not distracting.

ccressent commented on content/sensu-go/6.1/observability-pipeline/observe-schedule/subscriptions.md:
```diff
@@ -0,0 +1,158 @@
+    "entity:i-828282"
+  ]
+}
+{{< /code >}}
```
This is great, I didn't even know one could do that!
(I haven't tried it, I'm assuming you have)

hillaryfraley replied:
It is a good feature! I have tried and confirmed it on my local setup.

ccressent commented on content/sensu-go/6.1/observability-pipeline/observe-schedule/subscriptions.md:
```diff
@@ -0,0 +1,158 @@
+sensu-agent start --subscriptions linux,webserver --log-level debug
+{{< /code >}}
+
+Now your agent will execute checks with the `linux` or `webserver` subscriptions.
```
Suggestion: mention again and link to other methods to add/remove subscriptions to entities.

ccressent reviewed (approved):
Looks great, very easy to understand in my opinion.
I've left a few suggestions, but it's good to go as it stands.

hillaryfraley committed:
Updates from Cyril's review

hillaryfraley committed:
Propagate to 6.0 docs

hillaryfraley committed:
Few updates to schedule index page
